### DISCUSSION TITLE
Storing Directus metadata in a different database and keeping user database clean

### DISCUSSION THREAD

**Discussion by shellking4** [directus/directus#19297] [Feature Requests]
### Summary

Directus is already quite a great product. I must say that the way things are being done is really great compared to some alternatives solutions. But as we all know there is always room for enhancements and improvements. I was reading through nocodb's architecture and saw this. https://docs.nocodb.com/engineering/architecture. 
The arechitecture looks pretty much like the one of Directus except that they keep metadata in a different database. If user doesn't specify anything for that metadata database at the start of the project, they just store everything metadata related in a sqlite database. And afterwards the user can connect to his database. And that one is just introspected, nothing is done in it, no table created. I think it would be really great to adopt that kind of structure in Directus. So that  **directus_**  tables will be stored in another database (the directus metadata database, preferably sqlite). And the user-defined tables would stay clear in another database.  It could make things somewhat more fun to work with.

### Motivation

The purpose of this is to ease migrations and remove every risk for a user-defined table deletion when migrating between two Directus projects or when migrating from Directus

### Detailed Design

https://drive.google.com/file/d/1eroV6Fnc3pwKSfpf-DQe8DbhdczM63Uc/view?usp=sharing

### Requirements List

https://drive.google.com/file/d/1eroV6Fnc3pwKSfpf-DQe8DbhdczM63Uc/view?usp=sharing

### Drawbacks

Can be a little more complex than the current architecture

### Alternatives

Keep the current architecture

### Adoption Strategy

It would be a little bit hard to use for old Directus project but It would be a little win for new projects

**Comment by DanielBiegler** [comment]
I think in order to make large architectural changes viable, the reward (i.e. performance, ease of use, features, etc.) need to be large too to justify the undertaking and at least from my understanding right now, I don't see splitting up the DB to be useful enough? Maybe someone with a deeper understanding can enlighten me here but so far I dont see it.

> It could make things somewhat more fun to work with.

How so?

**Reply by shellking4** [reply]
Yeah I also think it would lead to some more works and possibly some performance issue too. But as I said the win will be that the user database will be more clear, containing only business models tables created by the end developer

**Comment by Zyles** [comment]
Storing the data in sqlite is bad once you start scaling your application horizontally. Each container would have their own sqlite database. Sqlite is only usable in very small projects or when developing locally.

And also by doing this you actually increase the complexity of the project not decrease it.

**Reply by shellking4** [reply]
Ok I got that SQLite maybe a bad choice
But SQLite was just an example I gave for sake of simplicity. In the nocodb architecture which inspires me the idea in the first place, it is possible to use something more robust like Postgres for the metadata database if the user configures it to be so

**Comment by bernatvadell** [comment]
From my point of view, this task would be the predecessor to a much more ambitious task, multiple data sources.

Keeping the schema separate would allow adding not one, but several alternative data sources. With some effort, even Directus could simplify the crossing of information, just as Hasura does.

This change would be a big step forward, I imagine Directus bringing order to larger projects that manage multiple databases. ❤️
